File: Programacao_Logica/05_recursao/codes/fibo.pl
% Computa o n-ésimo número de Fibonacci
fibTR(N, B, A, F) :-
    N > 0,
    succ(NewN, N),
    C is A + B,
    fibTR(NewN, C, B, F).

fibTR(0, _, Acc, Acc).

:- use_module(library(arithmetic)).
:- arithmetic_function(fib/1).

fib(N, F) :- fibTR(N, 1, 0, F).

% Exemplo de consulta. SWI-Prolog implementa aritmética estendida
?- X is fib(200).
X = 280571172992510140037611932413038677189525.

File: Programacao_Logica/05_recursao/codes/fibo.py
# Implementação iterativa
def fib(N):
    if N < 2:
        return N

    a = 0
    b = 1

    for _ in range(N):
        a, b = b, a + b

    return a

# Computa o bilionésimo número de Fibonacci em Python
# Tempo de execução: 9,251s
print(fib(10**6))
